### PR TITLE
GVT-3046: Suunnitelmien aluehaun bugikorjauksia ja yleistä refaktorointia

### PR DESCRIPTION
**Pull request by juuseko**
Varsinaisena tiketille kirjattuna bugina oli, että haun saa heittämään tietyissä tilanteissa internal server erroria. Tämä korjattu, ja samalla hakua on palasteltu ja muokattu muutenkin fiksumpaan muotoon.

### CODE REVIEW

File: infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutAlignmentService.kt [hunk 1]
```diff
@@ -84,46 +86,54 @@ class LayoutAlignmentService(
...
-            )
-
-        val polygon = bufferedPolygonForLineStringPoints(simplified, ALIGNMENT_POLYGON_BUFFER, LAYOUT_SRID)
-        val plans = geometryDao.fetchIntersectingPlans(polygon, LAYOUT_SRID)
-        return geometryDao.getPlanHeaders(plans)
+        val alignment = layoutAlignmentDao.fetch(alignmentVersion)
+        val geocodingContext = requireNotNull(geocodingService.getGeocodingContext(geocodingContextCacheKey))
+        if (!cropKmsAreWithinGeocodingContext(startKmNumber, endKmNumber, geocodingContext)) {
+            return emptyList()
+        }
```
Comment (juuseko) [inline infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutAlignmentService.kt:97]: Aiemmin meillä kävi niin, että jos operaattorin syöttämät ratakilometrit osuivat kokonaan koko alignmentin geokoodauskontekstin ulkopuolelle, niin palautettiin kaikki suunnitelmat. Esimerkki tällaisesta olisi esim. jos operaattori syötti haetuksi raiteeksi sellaisen raiteen jonka pituusmittauslinja kulkee välillä 0100-0200 ja hakuväliksi kilometrit 0010-0050. Tämä tarkastus tehdään erillään ja puhtaasti kilometrinumeroiden perusteella ennen varsinaista M-arvojen hakua, sillä M-arvot on kuitenkin määritelty vain ratanumeron kilometrialueen sisällä

File: infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutAlignmentService.kt [hunk 2]
```diff
@@ -84,46 +86,54 @@ class LayoutAlignmentService(
...
+
+        val (cropStartM, cropEndM) = getMValuesAlongReferenceLineForCrop(geocodingContext, startKmNumber, endKmNumber)
+        if (!alignmentOverlapsCropRange(alignment, geocodingContext, cropStartM, cropEndM)) {
+            return emptyList()
+        } else {
```
Comment (juuseko) [inline infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutAlignmentService.kt:102]: Tässä toinen lisätarkastelu, joka on validi lähinnä sijaintiraiteille. Jos annettu kilometriväli osuu annetun pituusmittauslinjan sisälle, niin tehdään tarkempi, M-arvopohjainen tarkastelu osuuko annettu kilometriväli raiteen sisälle

File: infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutAlignmentService.kt [hunk 3]
```diff
@@ -82,64 +76,51 @@ class LayoutAlignmentService(
...
-                    geocodingContext?.referencePoints?.find { it.kmNumber == startKmNumber }?.distance
-                } ?: alignment.start?.m
-            )
-        val endM =
-            requireNotNull(
-                endKmNumber?.let { geocodingContext?.referencePoints?.find { it.kmNumber > endKmNumber }?.distance }
-                    ?: alignment.end?.m
-            )
-        val mRange = Range(startM, endM)
+    private fun cropAlignment(alignment: LayoutAlignment, cropStartM: Double?, cropEndM: Double?): IAlignment {
```
Comment (juuseko) [inline infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutAlignmentService.kt:100]: Alignmentin croppaaminen on muuttunut nyt merkittävästi simppelimmäksi. Enää ei tarvita geokoodauskontekstia että saataisiin tehtyä (ei edes kauhean toimivia) validointeja sille osuvatko annetut ratakilometrit ylipäätään alignmentin alueelle, vaan voidaan vaan filtteröidä segmenttejä M-arvoilla

File: infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutAlignmentService.kt [hunk 4]
```diff
@@ -160,3 +141,24 @@ private fun toPlanSectionPoint(point: IPoint, alignment: LayoutAlignment, contex
...
+        geocodingContext.referencePoints.first().kmNumber <= endKmNumber
+    } else if (startKmNumber != null && endKmNumber != null && startKmNumber <= endKmNumber) {
+        val kmNumberRange = Range(startKmNumber, endKmNumber)
+        val geocodingRange =
+            Range(geocodingContext.referencePoints.first().kmNumber, geocodingContext.referencePoints.last().kmNumber)
+        kmNumberRange.overlaps(geocodingRange) || geocodingRange.overlaps(kmNumberRange)
+    } else {
+        startKmNumber == null && endKmNumber == null
+    }
+}
```
Comment (juuseko) [inline infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutAlignmentService.kt:164]: Eriytetty omakseen `cropAlignment()`:ista. Funktio on sen lisäksi korjattu, sillä alkuperäinen vertailu toimi tietyissä tilanteissa väärin. Jätetty asumaan tänne, sillä tämän tarkastelua tarvitaan sekä ratanumeroille että sijaintiraiteille

File: infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutTrackNumberService.kt [hunk 5]
```diff
@@ -204,14 +205,32 @@ class LayoutTrackNumberService(
...
+            )
+        } else {
+            val (cropStartM, cropEndM) = getCropMValues(geocodingContext, startKmNumber, endKmNumber)
+            return alignmentService.getOverlappingPlanHeaders(
+                alignment = layoutAlignmentDao.fetch(alignmentVersion),
+                polygonBufferSize = polygonBufferSize,
+                cropStartM = cropStartM,
+                cropEndM = cropEndM,
+            )
+        }
```
Comment (juuseko) [inline infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutTrackNumberService.kt:233]: Aiemman suoraan aina päätyyn saakka menemisen sijasta tänne on toteutettu vähän sisäistä logiikkaa:
1. Tarkastellaan sanity checkinä että jos kilometriväli on annettu, niin osuuko se ylipäätään pituusmittauslinjan alueelle ja jos ei, palautetaan tyhjä lista
2. Jos croppausaluetta ei ole annettu niin sen M-arvoissa ei ole mitään laskettavaa -> oikosuljetaan tapaus pois
3. Jos croppausalue on annettu ja se osuu pituusmittauslinjan alueelle, niin lasketaan alueelle M-arvot vasta silloin

File: infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LocationTrackService.kt [hunk 6]
```diff
@@ -392,15 +392,66 @@ class LocationTrackService(
...
+                alignment,
+                context,
+                polygonBufferSize,
+                trackStart,
+                trackEnd,
+                cropStart,
+                cropEnd,
+            )
+        }
+    }
```
Comment (juuseko) [inline infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LocationTrackService.kt:422]: Sijaintiraiteille toteutettu kanssa vastaavaa logiikkaa kuin ratanumeroille. Täällä asiat eivät kuitenkaan mene ihan yhtä simppelisti, sillä puhtaan ratakilometrien geokoodauskontekstiin vertailun lisäksi pitää sijaintiraiteilla vertailla myös M-arvoja. Tämä funktio tekee käytännössä samat vertailut ja oikosulkemiset kuin ratanumeroiden vastaava. Suurin ero tulee siitä, että M-arvopohjainen vertailu on erotettu omaan funktioonsa (alla)

File: infra/src/test/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutAlignmentServiceIT.kt
Comment (juuseko) [inline infra/src/test/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutAlignmentServiceIT.kt:1]: Testit siirtyivät täältä erillisiin ratanumero- ja sijaintiraidetesteihin

File: infra/src/test/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutTrackNumberServiceIT.kt [hunk 7]
```diff
@@ -407,6 +419,338 @@ constructor(
...
     }
 
+    @Test
+    fun `overlapping plan search finds plans that are within 10m of alignment`() {
+        val tn = TrackNumber("001")
+        // TODO: The alignments with "_2" in their names exist because of a bug that causes bounding polygons for plans
+        // with alignments in a straight line to be treated as linestrings and throw an SQL error when trying to
+        // be saved into the database. This a workaround to make the bounding polygon a valid polygon instead of a
+        // linestring. Remove them after GVT-3040 has been done
+        val a1_1 = geometryAlignment(line(Point(0.0, 0.0), Point(10.0, 0.0)))
```
Comment (juuseko) [inline infra/src/test/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutTrackNumberServiceIT.kt:429]: Saisipa nämä detektivaroitukset käännettyä jotenkin massana pois :thiswasfine: Näiden korjaaminen kommentin mukaanhan tulee mahdolliseksi kunhan Ari saa tikettinsä mainiin

File: infra/src/test/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutTrackNumberServiceIT.kt [hunk 8]
```diff
@@ -407,6 +419,338 @@ constructor(
...
+
+        assertEquals(4, overlapping.size)
+        assertContains(overlapping, plan2EndsWithinAlignmentBuffer.id)
+        assertContains(overlapping, plan3CompletelyWithin.id)
+        assertContains(overlapping, plan4TouchesEndOfAlignmentBuffer.id)
+        assertContains(overlapping, plan7TouchesBufferFromSide.id)
+    }
+
+    @Test
+    fun `overlapping plan search cropping works correctly in a happy case`() {
```
Comment (juuseko) [inline infra/src/test/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutTrackNumberServiceIT.kt:517]: Tässä kanssa wanha testi, joka vain siirretty ja muutettu pelaamaan ratanumerolla eikä sijaintiraiteella

File: infra/src/test/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutTrackNumberServiceIT.kt [hunk 9]
```diff
@@ -407,6 +419,338 @@ constructor(
...
+                .map { it.id }
+        assertEquals(4, overlapping.size)
+        assertContains(overlapping, plan2EndsBeforeStartKmButWithinBuffer.id)
+        assertContains(overlapping, plan3IsCompletelyWithinKmRange.id)
+        assertContains(overlapping, plan4StartsWithinKmRangeButEndsAfter.id)
+        assertContains(overlapping, plan5TouchesEndKmWhenBufferIsIncluded.id)
+    }
+
+    @Test
+    fun `overlapping plan search cropping works correctly in different edge cases`() {
```
Comment (juuseko) [inline infra/src/test/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutTrackNumberServiceIT.kt:641]: Uusi testi

File: infra/src/test/kotlin/fi/fta/geoviite/infra/tracklayout/LocationTrackServiceIT.kt
Comment (juuseko) [inline infra/src/test/kotlin/fi/fta/geoviite/infra/tracklayout/LocationTrackServiceIT.kt:1]: Tänne kanssa tehty samat muutokset kuin ylempään tiedostoon

File: infra/src/test/kotlin/fi/fta/geoviite/infra/tracklayout/LocationTrackServiceIT.kt [hunk 10]
```diff
@@ -787,6 +799,369 @@ constructor(
...
+                .map { it.id }
+        assertEquals(4, overlapping.size)
+        assertContains(overlapping, plan2EndsBeforeStartKmButWithinBuffer.id)
+        assertContains(overlapping, plan3IsCompletelyWithinKmRange.id)
+        assertContains(overlapping, plan4StartsWithinKmRangeButEndsAfter.id)
+        assertContains(overlapping, plan5TouchesEndKmWhenBufferIsIncluded.id)
+    }
+
+    @Test
+    fun `overlapping plan search cropping works correctly in different edge cases`() {
```
Comment (juuseko) [inline infra/src/test/kotlin/fi/fta/geoviite/infra/tracklayout/LocationTrackServiceIT.kt:1028]: Tässä kanssa uusi testi. Tämä on monimutkaisempi kuin ratanumeron vastaava, pääasiassa johtuen kaksitasoisesta validoinnista

**Review by solita-mrunsten — COMMENTED** [review]
👍

**Review by jyrkija — COMMENTED** [review]
:+1: 
En käynyt nyt kovin tarkkaan läpi koska tämä tuli aika isosti vastaan graafi-puolelle mergeillessä (koska siellä raiteen geometrian rakenne on niin eri). Siellä tätä on nyt jonkun verran muutettu taas toiseen asentoon, joten kannattanee välttää muuttelemasta isommin mainin puolella kunnes saadaan (toivottavasti pian) ne mainiin.
